src/datachain/lib/signal_schema.py

```diff
@@ -749,3 +754,108 @@ def _build_tree_for_model(
+        custom_types = serialized.get("_custom_types", {})
+
+        schema: dict[str, Any] = {}
+        schema_custom_types: dict[str, CustomType] = {}
+
+        data_model_bases: Optional[list[tuple[str, str, Optional[str]]]] = None
+
+        for column in columns:
+            parent_type_origin, parent_type_partial = "", ""
+            for i, signal in enumerate(column.split(".")):
+                if i == 0:
+                    if signal not in serialized:
+                        raise SignalSchemaError(
+                            f"Column {column} not found in the schema"
+                        )
+
+                    parent_type_origin = serialized[signal]
+                    parent_type_partial = _type_name_to_partial(parent_type_origin)
+
+                    schema[signal] = parent_type_partial
+                    continue
+
+                if parent_type_origin not in custom_types:
+                    raise SignalSchemaError(
+                        f"Custom type {parent_type_origin} not found in the schema"
+                    )
+
+                custom_type = custom_types[parent_type_origin]
+                signal_type = custom_type["fields"].get(signal)
+                if not signal_type:
+                    raise SignalSchemaError(
+                        f"Field {signal} not found in custom type {parent_type_origin}"
+                    )
+
+                partial_type = _type_name_to_partial(signal_type)
+
+                if parent_type_partial in schema_custom_types:
+                    schema_custom_types[parent_type_partial].fields[signal] = (
+                        partial_type
+                    )
+                else:
+                    if data_model_bases is None:
+                        data_model_bases = SignalSchema._get_bases(DataModel)
+
+                    partial_type_name, _ = ModelStore.parse_name_version(partial_type)
+                    schema_custom_types[parent_type_partial] = CustomType(
+                        schema_version=2,
+                        name=partial_type_name,
+                        fields={signal: partial_type},
+                        bases=[
+                            (partial_type_name, "__main__", partial_type),
+                            *data_model_bases,
+                        ],
+                    )
+
+                parent_type_origin, parent_type_partial = signal_type, partial_type
+
+        if schema_custom_types:
+            schema["_custom_types"] = {
+                type_name: ct.model_dump()
+                for type_name, ct in schema_custom_types.items()
+            }
+
+        return SignalSchema.deserialize(schema)
+
+
+def _type_name_to_partial(type_name: str) -> str:
+    """Convert a type name to a partial type name with the same version."""
+    if "@" not in type_name:
+        return type_name
+    model_name, version = ModelStore.parse_name_version(type_name)
+    return f"{model_name}Partial@v{version}"
```

Comment:
can we hit to different Partial constructed from the same model with a bit different set of fields? (this function will generate the same name for both which might be wrong)

Reply:
I think it is possible in theory, but I can not think of any valid real-world example. Partial schema is created from existing dataset and if dataset's schema is correct, partial schema will also be correct.

Same is possible in theory, for example, if we will use `merge` from different datasets or `map` from different Python files with different custom models with the same name.

Not sure if we need to address this issue in this PR, but this is a good question for sure and we should think about it.

Reply:
Two files (File model). Different name in the DB schema:

`file1`: File
`file2`: File

we use columns `file1.path` and `file2.source`.

will it lead to a collision here potentially?

Reply:
One partial model will be created:

```
FilePartial@v1
    path: str
    source: str
```

Reply:
kk, is it correct though? does it mean we'll keep some extra fields?